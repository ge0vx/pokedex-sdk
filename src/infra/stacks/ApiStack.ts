import { Stack, StackProps } from "aws-cdk-lib";
import { Cors, LambdaIntegration, ResourceOptions, RestApi } from "aws-cdk-lib/aws-apigateway";
import { Construct } from "constructs";

interface ApiStackProps extends StackProps {
    favPokemonIntegration: LambdaIntegration
}

export class ApiStack extends Stack {

    constructor(scope: Construct, id: string, props?: ApiStackProps){
        super(scope, id, props);

        const api = new RestApi(this, 'favPokemonApi');

        const optionsWithCors: ResourceOptions = {
            defaultCorsPreflightOptions: {
                allowOrigins: Cors.ALL_ORIGINS,
                allowMethods: Cors.ALL_METHODS,
            }
        }

        const favPokemonResource = api.root.addResource('favPokemon', optionsWithCors);
        favPokemonResource.addMethod('GET', props.favPokemonIntegration)
        favPokemonResource.addMethod('POST', props.favPokemonIntegration)
        
    }
}