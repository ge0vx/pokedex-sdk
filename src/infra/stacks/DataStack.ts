import { Stack, StackProps } from "aws-cdk-lib";
import { AttributeType, ITable, Table } from "aws-cdk-lib/aws-dynamodb";
import { Construct } from "constructs";

export class DataStack extends Stack {

    public readonly favPokemonTable: ITable;

    constructor(scope: Construct, id: string, props?: StackProps){
        super(scope, id, props)

        this.favPokemonTable = new Table(this, 'favPokemonTable', {
            partitionKey : {
                name: 'id',
                type: AttributeType.STRING
            },
            tableName : 'favPokemonTable'
        })
    }
}