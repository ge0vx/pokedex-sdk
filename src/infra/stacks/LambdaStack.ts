import { Stack, StackProps } from "aws-cdk-lib";
import { LambdaIntegration } from "aws-cdk-lib/aws-apigateway";
import { ITable } from "aws-cdk-lib/aws-dynamodb";
import { Runtime } from "aws-cdk-lib/aws-lambda";
import { NodejsFunction } from "aws-cdk-lib/aws-lambda-nodejs";
import { Construct } from "constructs";
import { join } from "path";
import { Effect, PolicyStatement } from "aws-cdk-lib/aws-iam";

interface LambdaStackProps extends StackProps {
    favPokemonTable: ITable
}

export class LambdaStack extends Stack {

    public readonly favPokemonIntegration: LambdaIntegration;
    public readonly pokemonLamdaIntegration: LambdaIntegration;

    constructor(scope: Construct, id: string, props?: LambdaStackProps){
        super(scope, id, props);

        const runtime = Runtime.NODEJS_18_X;
        const handlerFileName = 'handler';

        const favPokemonLambda = new NodejsFunction(this, 'favPokemonLambda',{
            runtime,
            handler: handlerFileName,
            entry: join(__dirname, '..', '..', 'services', 'favPokemon', 'handler.ts'),
            environment: {
                TABLE_NAME: props.favPokemonTable.tableName
            }
        })

        favPokemonLambda.addToRolePolicy(new PolicyStatement({
            effect: Effect.ALLOW,
            resources: [props.favPokemonTable.tableArn],
            actions:[
                'dynamodb:PutItem',
                'dynamodb:Scan'
            ]
        }))

        const pokemonLambda = new NodejsFunction(this, 'PokemonLamda', {
            runtime: runtime,
            handler: handlerFileName,
            entry: join(__dirname, '..', '..', 'services', 'pokemon', 'handler.ts'),
        })
        
        this.favPokemonIntegration = new LambdaIntegration(favPokemonLambda)
        this.pokemonLamdaIntegration = new LambdaIntegration(pokemonLambda)
    }
}