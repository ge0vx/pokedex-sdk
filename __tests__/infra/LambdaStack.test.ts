import { App } from "aws-cdk-lib";
import { LambdaStack } from "../../src/infra/stacks/LambdaStack";
import { Template } from "aws-cdk-lib/assertions";
import { DataStack } from "../../src/infra/stacks/DataStack";

describe('Initial test suite', ()=>{

    let lambdaStackTemplate: Template;

    beforeAll(()=>{
        const testApp = new App({
            outdir: 'cdk.out'
        })

        const dataStack = new DataStack(testApp, 'DataStackFavPokemon');
        const lambdaStack = new LambdaStack(testApp, 'LambdaStackFavPokemon', {
            favPokemonTable : dataStack.favPokemonTable
        });

        lambdaStackTemplate = Template.fromStack(lambdaStack);
    })


    test('lambda properties', ()=>{
        lambdaStackTemplate.hasResourceProperties('AWS::Lambda::Function', {
            Handler: "index.handler",
            Runtime: "nodejs18.x"
        })
    });

    test('IAM Role snapshot', ()=>{
        const iamRole = lambdaStackTemplate.findResources('AWS::IAM::Role')
        expect(iamRole).toMatchSnapshot();
    });

    test('IAM Policy snapshot', ()=>{
        const iamPolicy = lambdaStackTemplate.findResources('AWS::IAM::Role')
        expect(iamPolicy).toMatchSnapshot();
    });

    test('Lambda service snapshot', ()=>{
        const lambda = lambdaStackTemplate.findResources('AWS::Lambda::Function')
        expect(lambda).toMatchSnapshot();
    });
});