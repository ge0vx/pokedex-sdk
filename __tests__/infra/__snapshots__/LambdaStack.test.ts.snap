// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Initial test suite IAM Policy snapshot 1`] = `
{
  "favPokemonLambdaServiceRole92C59608": {
    "Properties": {
      "AssumeRolePolicyDocument": {
        "Statement": [
          {
            "Action": "sts:AssumeRole",
            "Effect": "Allow",
            "Principal": {
              "Service": "lambda.amazonaws.com",
            },
          },
        ],
        "Version": "2012-10-17",
      },
      "ManagedPolicyArns": [
        {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
            ],
          ],
        },
      ],
    },
    "Type": "AWS::IAM::Role",
  },
}
`;

exports[`Initial test suite IAM Role snapshot 1`] = `
{
  "favPokemonLambdaServiceRole92C59608": {
    "Properties": {
      "AssumeRolePolicyDocument": {
        "Statement": [
          {
            "Action": "sts:AssumeRole",
            "Effect": "Allow",
            "Principal": {
              "Service": "lambda.amazonaws.com",
            },
          },
        ],
        "Version": "2012-10-17",
      },
      "ManagedPolicyArns": [
        {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
            ],
          ],
        },
      ],
    },
    "Type": "AWS::IAM::Role",
  },
}
`;

exports[`Initial test suite Lambda service snapshot 1`] = `
{
  "favPokemonLambdaBFB7282A": {
    "DependsOn": [
      "favPokemonLambdaServiceRoleDefaultPolicy6420EA11",
      "favPokemonLambdaServiceRole92C59608",
    ],
    "Properties": {
      "Code": {
        "S3Bucket": {
          "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
        },
        "S3Key": "d35672eb5ceacf091e7e17b942f7cc8f0dfa61361c5344dd7fa553c7c1ebb6b7.zip",
      },
      "Environment": {
        "Variables": {
          "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
          "TABLE_NAME": {
            "Fn::ImportValue": "DataStackFavPokemon:ExportsOutputReffavPokemonTable9241278D998B3767",
          },
        },
      },
      "Handler": "index.handler",
      "Role": {
        "Fn::GetAtt": [
          "favPokemonLambdaServiceRole92C59608",
          "Arn",
        ],
      },
      "Runtime": "nodejs18.x",
    },
    "Type": "AWS::Lambda::Function",
  },
}
`;
